notebooks/optuna-example.ipynb

```diff
@@ -0,0 +1,495 @@
+    "    y = y.astype(np.int64)\n",
+    "    return X, y"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MyModule(nn.Module):\n",
+    "    def __init__(self, num_units=10, nonlin=nn.ReLU(), dropout_rate=0.5):\n",
+    "        super(MyModule, self).__init__()\n",
+    "        self.num_units = num_units\n",
+    "        self.nonlin = nonlin\n",
+    "        self.dropout_rate = dropout_rate\n",
+    "        \n",
+    "        self.reset_weights()\n",
+    "        \n",
+    "    def reset_weights(self):\n",
+    "        self.dense0 = nn.Linear(20, self.num_units)\n",
+    "        self.dropout = nn.Dropout(self.dropout_rate)\n",
+    "        self.dense1 = nn.Linear(self.num_units, self.num_units)\n",
+    "        self.output = nn.Linear(self.num_units, 2)\n",
+    "        self.softmax = nn.Softmax(dim=-1)\n",
+    "\n",
+    "    def forward(self, X, **kwargs):\n",
+    "        X = self.nonlin(self.dense0(X))\n",
+    "        X = self.dropout(X)\n",
+    "        X = self.nonlin(self.dense1(X))\n",
+    "        X = self.softmax(self.output(X))\n",
+    "        return X"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def define_model(trial):\n",
+    "    \"\"\"Model definition and hyper-parameter setting\n",
+    "\n",
+    "    We set the number of units, the optimizer, the learning rate, and the number of epochs.\n",
+    "    \n",
+    "    To make this work efficiently with optuna, add the\n",
+    "    `SkorchPruningCallback <https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html?highlight=optuna.integration.OptunaSearchCV>`\n",
+    "    from the optuna library.\n",
+    "\n",
+    "    \"\"\"\n",
+    "    num_units = trial.suggest_int('num_units', 10, 50)\n",
+    "    optimizer_name = trial.suggest_categorical('optimizer', ['SGD', 'Adam'])\n",
+    "    optimizer = getattr(optim, optimizer_name)\n",
+    "    lr = trial.suggest_float('lr', 1e-5, 1e-1, log=True)\n",
+    "    max_epochs = trial.suggest_int('max_epochs', 5, 25)\n",
+    "    \n",
+    "    return NeuralNetClassifier(\n",
+    "        MyModule,\n",
+    "        module__num_units=num_units,\n",
+    "        optimizer=optimizer,\n",
+    "        lr=lr,\n",
+    "        max_epochs=max_epochs,\n",
+    "        verbose=0,\n",
+    "        device=DEVICE,\n",
+    "        callbacks=[SkorchPruningCallback(trial, monitor='valid_acc')],\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def objective(trial):\n",
+    "    \"\"\"Train the model and return the value to be maximized\"\"\"\n",
+    "    model = define_model(trial)\n",
+    "    X, y = get_data()\n",
+    "    model.fit(X, y)\n",
+    "    accuracy = model.history[-1, METRIC_NAME]\n",
```

Comment:
Would it be cleaner to directly place the metric name here?

```suggestion
    "    accuracy = model.history[-1, 'valid_acc']\n",
```

Reply:
I would leave it as is. The idea is to set all the constants at the beginning of the notebook. This way, if I want to change the them (say, optimizing `valid_loss` instead), I don't need to chase through the notebook to replace all instances.